src/api/bulk2.ts

```diff
@@ -0,0 +1,1082 @@
+
+type CreateJobDataV2Options<S extends Schema> = {
+  job: IngestJobV2<S>;
+  createRequest: CreateIngestJobV2Request;
+};
+
+type CreateQueryJobV2Options = {
+  bodyParams: NewQueryJobOptions;
+  pollingOptions: BulkV2PollingOptions;
+};
+
+type ExistingQueryJobV2Options = {
+  id: string;
+  pollingOptions: BulkV2PollingOptions;
+};
+
+type BulkV2PollingOptions = {
+  pollInterval: number;
+  pollTimeout: number;
+};
+
+class JobPollingTimeoutError extends Error {
+  jobId: string;
+
+  /**
+   *
+   */
+  constructor(message: string, jobId: string) {
+    super(message);
+    this.name = 'JobPollingTimeout';
+    this.jobId = jobId;
+  }
+}
+
+class BulkApiV2<S extends Schema> extends HttpApi<S> {
+  hasErrorInResponseBody(body: any) {
+    return (
+      Array.isArray(body) &&
+      typeof body[0] === 'object' &&
+      'errorCode' in body[0]
+    );
+  }
+
+  isSessionExpired(response: HttpResponse): boolean {
+    return (
+      response.statusCode === 401 && /INVALID_SESSION_ID/.test(response.body)
+    );
+  }
+
+  parseError(body: any) {
+    return {
+      errorCode: body[0].errorCode,
+      message: body[0].message,
+    };
+  }
+}
+
+export class BulkV2<S extends Schema> {
+  private connection: Connection<S>;
+  private logger: Logger;
+
+  /**
+   * Polling interval in milliseconds
+   *
+   * Default: 1000 (1 second)
+   */
+  public pollInterval = 1000;
+
+  /**
+   * Polling timeout in milliseconds
+   *
+   * Default: 30000 (30 seconds)
+   */
+  public pollTimeout = 30000;
+
+  constructor(connection: Connection<S>) {
+    this.connection = connection;
+    this.logger = this.connection._logLevel
+      ? getLogger('bulk2').createInstance(this.connection._logLevel)
+      : getLogger('bulk2');
+  }
+
+  /**
+   * Create an instance of an ingest job object.
+   *
+   * @params {NewIngestJobOptions} options object
+   * @returns {IngestJobV2} An ingest job instance
+   * @example
+   * // Upsert records to the Account object.
+   *
+   * const job = connection.bulk2.createJob({
+   *   operation: 'insert'
+   *   object: 'Account',
+   * });
+   *
+   * // create the job in the org
+   * await job.open()
+   *
+   * // upload data
+   * await job.uploadData(csvFile)
+   *
+   * // finished uploading data, mark it as ready for processing
+   * await job.close()
+   */
+  public createJob(options: NewIngestJobOptions): IngestJobV2<S> {
+    return new IngestJobV2(this.connection, {
+      bodyParams: options,
+      pollingOptions: this,
+    });
+  }
+
+  /**
+   * Get an ingest or query job instance specified by a given job ID
+   *
+   * @param options Options object with a job ID
+   * @returns IngestJobV2 An ingest job
+   */
+  public job(
+    type: 'query',
+    options: Pick<ExistingQueryJobV2Options, 'id'>,
+  ): QueryJobV2<S>;
+  public job(
+    type: 'ingest',
+    options: Pick<ExistingIngestJobOptions, 'id'>,
+  ): IngestJobV2<S>;
+  public job(
```

Comment:
get an existing job by providing an id.

by default will get an ingest job (this matches bulkv1.job), can get a query job too.